src/librustdoc/html/render.rs

```diff
@@ -2707,22 +2719,39 @@ fn render_impl(w: &mut fmt::Formatter, cx: &Context, i: &Impl, link: AssocItemLi
     }
 
     fn doc_impl_item(w: &mut fmt::Formatter, cx: &Context, item: &clean::Item,
-                     link: AssocItemLink, render_static: bool,
+                     link: AssocItemLink, render_mode: RenderMode,
                      is_default_item: bool, outer_version: Option<&str>,
                      trait_: Option<&clean::Trait>) -> fmt::Result {
         let item_type = item_type(item);
         let name = item.name.as_ref().unwrap();
 
-        let is_static = match item.inner {
-            clean::MethodItem(ref method) => !method.decl.has_self(),
-            clean::TyMethodItem(ref method) => !method.decl.has_self(),
-            _ => false
+        let render_method_item: bool = match render_mode {
+            RenderMode::Normal => true,
+            RenderMode::ForDeref { mut_: deref_mut_ } => {
+                let self_type_opt = match item.inner {
+                    clean::MethodItem(ref method) => method.decl.self_type(),
+                    clean::TyMethodItem(ref method) => method.decl.self_type(),
+                    _ => None
+                };
+
+                if let Some(self_ty) = self_type_opt {
+                    match self_ty {
+                        SelfTy::SelfValue => true,
+                        SelfTy::SelfBorrowed(_lifetime, mutability) => {
+                            deref_mut_ || mutability == Mutability::Immutable
+                        },
+                        SelfTy::SelfExplicit(_) => false,
```

Comment:
This also matches things like `self: &Self`, `self: &Foo`, `self: &mut Self`, `self: &mut Foo`, `self: Self` and `self: Foo` which we don't want to remove.